content/advent-2016/gleam-distributed-map-reduce-for-golang.md

```diff
@@ -0,0 +1,356 @@
+only agents with enough resources can get the job. So with this hint, 
+Gleam automatically allocates the memory.
+
+```go
+package main
+
+import (
+	"os"
+
+	"github.com/chrislusf/gleam/distributed"
+	"github.com/chrislusf/gleam/flow"
+)
+
+func main() {
+	flow.New().TextFile(
+	  "/Users/chris/Desktop/record_1Gb_input.txt",
+	).Hint(flow.TotalSize(1024)).Map(`
+	  function(line)
+	    return string.sub(line, 1, 10), string.sub(line, 13)
+	  end
+	`).Partition(4).Sort().Fprintf(os.Stdout, "%s  %s\n").Run(distributed.Option())
+}
+```
+Result:
+```sh
+$ time ./gleam_sort > result3.txt
+2016/11/25 17:54:42   taskGroup:Map.0-ScatterPartitions.0 : ScatterPartitions (5 MB)
+2016/11/25 17:54:42 localhost:45326 allocated 1 executors with 5 MB memory.
+2016/11/25 17:54:42   taskGroup:CollectPartitions.0-LocalSort.0 : CollectPartitions (3 MB)
+2016/11/25 17:54:42   taskGroup:CollectPartitions.0-LocalSort.0 : LocalSort (768 MB)
+2016/11/25 17:54:42 localhost:45326 allocated 1 executors with 771 MB memory.
+2016/11/25 17:54:42   taskGroup:CollectPartitions.2-LocalSort.2 : CollectPartitions (3 MB)
+2016/11/25 17:54:42   taskGroup:CollectPartitions.2-LocalSort.2 : LocalSort (768 MB)
+2016/11/25 17:54:42   taskGroup:CollectPartitions.1-LocalSort.1 : CollectPartitions (3 MB)
+2016/11/25 17:54:42   taskGroup:CollectPartitions.1-LocalSort.1 : LocalSort (768 MB)
+2016/11/25 17:54:42   taskGroup:CollectPartitions.3-LocalSort.3 : CollectPartitions (3 MB)
+2016/11/25 17:54:42   taskGroup:CollectPartitions.3-LocalSort.3 : LocalSort (768 MB)
+2016/11/25 17:54:42 localhost:45326 allocated 3 executors with 2313 MB memory.
+2016/11/25 17:54:42   taskGroup:MergeSortedTo.0 : MergeSortedTo (5 MB)
+2016/11/25 17:54:42 localhost:45326 allocated 1 executors with 5 MB memory.
+
+real	1m55.188s
+user	1m6.554s
+sys	0m18.362s
+
+```
+
+Each of the 4 sorting executors peaked at about 868 MB, at 100% CPU.
+
+The code used more CPU and memory than basic Unix sort. Instead of on-disk
+merge sort, the Gleam sort is just all in-memory Timsort.
+The actual sorting time is reduced from 4m57s to 1m55s, 39% of the baseline.
+
+# Example 2: Joining CSV files
+Data joining is another common use case for data processing. Here is a simple
+example to illustrate how it works.
+
+Assume there are file "a.csv" has fields "a1, a2, a3, a4, a5",
+and file "b.csv" has fields "b1, b2, b3". 
+
+We want to join the rows where a1 = b2. 
+
+And the output format should be "a1, a4, b3".
+
+```go
+package main
+
+import (
+    "os"
+
+    . "github.com/chrislusf/gleam/flow"
+    "github.com/chrislusf/gleam/plugins/csv"
+)
+
+func main() {
+
+    f := New()
+    a := f.ReadFile(csv.New("a.csv")).Select(Field(1,4)) // a1, a4
+    b := f.ReadFile(csv.New("b.csv")).Select(Field(2,3)) // b2, b3
+
+    a.Join(b).Fprintf(os.Stdout, "%s,%s,%s\n").Run()  // a1, a4, b3
+
+}
+
+```
+In the example, the csv files are on local disk. This is the simplest form. 
+The csv files can also be read from HDFS or S3.
```

Comment:
nit: I think `csv` should be `CSV`, this keeps consistent with `HDFS` and the next paragraph's use of `CSV`